### PR TITLE
[vectorized](function) support array_count function

### LINKED ISSUE #17313: [Feature] (array function) Support array_count function
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Returns the number of non-zero and non-null elements in the given array.

### Use case

```
mysql> select array_count(x -> x, [0, 1, 2, 3]);
+--------------------------------------------------------+
| array_count(array_map([x] -> x(0), ARRAY(0, 1, 2, 3))) |
+--------------------------------------------------------+
|                                                      3 |
+--------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select array_count(x -> x > 2, [0, 1, 2, 3]);
+------------------------------------------------------------+
| array_count(array_map([x] -> x(0) > 2, ARRAY(0, 1, 2, 3))) |
+------------------------------------------------------------+
|                                                          1 |
+------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select array_count(x -> x is null, [null, null, null, 1, 2]);
+----------------------------------------------------------------------------+
| array_count(array_map([x] -> x(0) IS NULL, ARRAY(NULL, NULL, NULL, 1, 2))) |
+----------------------------------------------------------------------------+
|                                                                          3 |
+----------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select array_count(x -> power(x,2)>10, [1, 2, 3, 4, 5]);
+------------------------------------------------------------------------------+
| array_count(array_map([x] -> power(x(0), 2.0) > 10.0, ARRAY(1, 2, 3, 4, 5))) |
+------------------------------------------------------------------------------+
|                                                                            2 |
+------------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select *, array_count((x, y) -> x>y, c_array1, c_array2) from array_test;
+------+-----------------+-------------------------+-----------------------------------------------------------------------+
| id   | c_array1        | c_array2                | array_count(array_map([x, y] -> x(0) > y(1), `c_array1`, `c_array2`)) |
+------+-----------------+-------------------------+-----------------------------------------------------------------------+
|    1 | [1, 2, 3, 4, 5] | [10, 20, -40, 80, -100] |                                                                     2 |
|    2 | [6, 7, 8]       | [10, 12, 13]            |                                                                     0 |
|    3 | [1]             | [-100]                  |                                                                     1 |
|    4 | [1, NULL, 2]    | [NULL, 3, 1]            |                                                                     1 |
|    5 | []              | []                      |                                                                     0 |
|    6 | NULL            | NULL                    |                                                                     0 |
+------+-----------------+-------------------------+-----------------------------------------------------------------------+
6 rows in set (0.02 sec)

```

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wzymumon**
# Proposed changes

Issue Number: close #17313 

## Problem summary

support array_count function.
array_count：Returns the number of non-zero and non-null elements in the given array.

```
mysql> select array_count(x -> x, [0, 1, 2, 3]);
+--------------------------------------------------------+
| array_count(array_map([x] -> x(0), ARRAY(0, 1, 2, 3))) |
+--------------------------------------------------------+
|                                                      3 |
+--------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select array_count(x -> x > 2, [0, 1, 2, 3]);
+------------------------------------------------------------+
| array_count(array_map([x] -> x(0) > 2, ARRAY(0, 1, 2, 3))) |
+------------------------------------------------------------+
|                                                          1 |
+------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select array_count(x -> x is null, [null, null, null, 1, 2]);
+----------------------------------------------------------------------------+
| array_count(array_map([x] -> x(0) IS NULL, ARRAY(NULL, NULL, NULL, 1, 2))) |
+----------------------------------------------------------------------------+
|                                                                          3 |
+----------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select array_count(x -> power(x,2)>10, [1, 2, 3, 4, 5]);
+------------------------------------------------------------------------------+
| array_count(array_map([x] -> power(x(0), 2.0) > 10.0, ARRAY(1, 2, 3, 4, 5))) |
+------------------------------------------------------------------------------+
|                                                                            2 |
+------------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select *, array_count((x, y) -> x>y, c_array1, c_array2) from array_test;
+------+-----------------+-------------------------+-----------------------------------------------------------------------+
| id   | c_array1        | c_array2                | array_count(array_map([x, y] -> x(0) > y(1), `c_array1`, `c_array2`)) |
+------+-----------------+-------------------------+-----------------------------------------------------------------------+
|    1 | [1, 2, 3, 4, 5] | [10, 20, -40, 80, -100] |                                                                     2 |
|    2 | [6, 7, 8]       | [10, 12, 13]            |                                                                     0 |
|    3 | [1]             | [-100]                  |                                                                     1 |
|    4 | [1, NULL, 2]    | [NULL, 3, 1]            |                                                                     1 |
|    5 | []              | []                      |                                                                     0 |
|    6 | NULL            | NULL                    |                                                                     0 |
+------+-----------------+-------------------------+-----------------------------------------------------------------------+
6 rows in set (0.02 sec)

```

## Checklist(Required)

* [x] Does it affect the original behavior
* [x] Has unit tests been added
* [x] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_count.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
+---
```
Comment (eldenmoon) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_count.md:1]: add sidebar to docs

File: docs/zh-CN/docs/sql-manual/sql-functions/array-functions/array_count.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+## array_count
+
+<version since="1.2.0">
```
Comment (eldenmoon) [inline docs/zh-CN/docs/sql-manual/sql-functions/array-functions/array_count.md:29]: change version to 1.2.4

File: be/src/vec/functions/array/function_array_count.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+#include <vec/columns/columns_number.h>
+#include <vec/data_types/data_type_array.h>
+#include <vec/data_types/data_type_number.h>
+#include <vec/functions/function.h>
+#include <vec/functions/function_helpers.h>
+#include <vec/functions/simple_function_factory.h>
+
+namespace doris::vectorized {
+
+// array_count([1, 2, 3, 0, 0]) -> 3
```
Comment (eldenmoon) [inline be/src/vec/functions/array/function_array_count.cpp:29]: we should support lambda function for the first argument

**Review by eldenmoon — CHANGES_REQUESTED** [review]

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_count.md (hunk 1 above)
Comment (wzymumon) [reply docs/en/docs/sql-manual/sql-functions/array-functions/array_count.md:1]: done

File: docs/zh-CN/docs/sql-manual/sql-functions/array-functions/array_count.md (hunk 2 above)
Comment (wzymumon) [reply docs/zh-CN/docs/sql-manual/sql-functions/array-functions/array_count.md]: done

**Comment by wzymumon** [comment]
> could you add one row like: [0,1,2,3,null] for test

done

File: be/src/vec/functions/array/function_array_count.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        DCHECK(is_array(arguments[0]))
+                << "first argument for function: " << name << " should be DataTypeArray";
+        return make_nullable(std::make_shared<DataTypeInt64>());
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) override {
+        // extract array offsets and nested data
+        auto left_column =
+                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
```
Comment (zclllyybb) [inline be/src/vec/functions/array/function_array_count.cpp:53]: Please dont use `convert_to_full_column_if_const`. Nowadays our common usage is listed in [document](https://selectdb.feishu.cn/docx/UtysdxKDkoJ5xoxtJEUcoN6ineU)

**Review by zclllyybb — CHANGES_REQUESTED** [review]

File: be/src/vec/functions/array/function_array_count.cpp (hunk 4 above)
Comment (wzymumon) [reply be/src/vec/functions/array/function_array_count.cpp:53]: > Please dont use `convert_to_full_column_if_const`. Nowadays our common usage is listed in [document](https://selectdb.feishu.cn/docx/UtysdxKDkoJ5xoxtJEUcoN6ineU)

ok. I will fix it later

File: be/src/vec/functions/array/function_array_count.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            if (array_null_map && array_null_map[i]) {
+                continue;
+            }
+            
+            // default count is 0 if no-zero elment is not found
+            Int64 count = 0;
+            for (size_t off = src_offset[i - 1]; off < src_offset[i]; ++off) {
+                if (!src_nested_data.is_null_at(off) && src_nested_data.get_bool(off)) {
+                    ++count;
```
Comment (eldenmoon) [inline be/src/vec/functions/array/function_array_count.cpp:78]: get_bool will call virtual function, and lead to performance issue, we should modify not call virtual function

Comment (eldenmoon) [reply be/src/vec/functions/array/function_array_count.cpp:78]: same as is_null_at

Comment (eldenmoon) [reply be/src/vec/functions/array/function_array_count.cpp:78]: we could do some cast here

File: be/src/vec/functions/array/function_array_count.cpp (hunk 4 above)
Comment (eldenmoon) [reply be/src/vec/functions/array/function_array_count.cpp:51]: remind

File: be/src/vec/functions/array/function_array_count.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+#include <vec/columns/columns_number.h>
+#include <vec/data_types/data_type_array.h>
+#include <vec/data_types/data_type_number.h>
+#include <vec/functions/function.h>
+#include <vec/functions/function_helpers.h>
+#include <vec/functions/simple_function_factory.h>
+
+namespace doris::vectorized {
+
+// array_count([1, 2, 3, 0, 0]) -> [3]
```
[the remark this answers is not in the document]
Comment (wzymumon) [reply be/src/vec/functions/array/function_array_count.cpp:29]: how about this? `array_count(x->x, [0, 1, 1, 1, 0]) -> [3]`

File: be/src/vec/functions/array/function_array_count.cpp (hunk 5 above)
Comment (wzymumon) [reply be/src/vec/functions/array/function_array_count.cpp]: done

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Comment by eldenmoon** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.08 seconds
 stream load tsv:          423 seconds loaded 74807831229 Bytes, about 168 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230508042353_clickbench_pr_139968.html

**Comment by wzymumon** [comment]
run buildall

**Comment by wzymumon** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by eldenmoon** [comment]
run buildall

**Review by eldenmoon — APPROVED** [review]
LGTM

**Review by eldenmoon — APPROVED** [review]
LGTM

**Review by eldenmoon — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**